File: packages/ruby-sdk/lib/strait/middleware.rb
# frozen_string_literal: true

module Strait
  # Middleware hooks for request/response/error lifecycle.
  Middleware = Struct.new(:on_request, :on_response, :on_error, keyword_init: true)

  # Context passed to on_request middleware hooks.
  MiddlewareRequestContext = Struct.new(:method, :url, :headers, keyword_init: true)

  # Context passed to on_response middleware hooks.
  MiddlewareResponseContext = Struct.new(:method, :url, :status, :duration_ms, keyword_init: true)

  # Context passed to on_error middleware hooks.
  MiddlewareErrorContext = Struct.new(:method, :url, :error, keyword_init: true)
end

File: packages/python-sdk/strait/_middleware.py
"""Middleware hooks for HTTP request/response interception."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass
class MiddlewareRequestContext:
    method: str
    url: str
    headers: dict[str, str]
    body: bytes | None = None


@dataclass
class MiddlewareResponseContext:
    method: str
    url: str
    status: int
    duration_ms: int


@dataclass
class MiddlewareErrorContext:
    method: str
    url: str
    error: BaseException


@dataclass
class Middleware:
    on_request: Callable[[MiddlewareRequestContext], None] | None = None
    on_response: Callable[[MiddlewareResponseContext], None] | None = None
    on_error: Callable[[MiddlewareErrorContext], None] | None = None
